oci-image/src/state.rs

```diff
@@ -0,0 +1,216 @@
+    fn has_layer(&self, layer_id: &str) -> bool;
+    /// Check if the state contains a snapshot
+    fn has_snapshot(&self, snapshot_id: &str) -> bool;
+    /// Get an image from the state by it's identifier
+    fn get_image(&self, image_id: &str) -> Option<&ImageMetadata>;
+    /// Get a layer from the state by it's identifier
+    fn get_layer(&self, layer_id: &str) -> Option<&LayerMetadata>;
+    // Get a snapshot from the state by it's identifier
+    fn get_snapshot(&self, snapshot_id: &str) -> Option<&usize>;
+}
+
+/// `LayerMetadata` struct holds information's about a layer in the state.
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct LayerMetadata {
+    /// The id of the layer
+    pub id: String,
+    /// The compressed layer digest
+    pub compressed_digest: String,
+    /// The decompressed layer digest,
+    pub uncompressed_digest: String,
+    /// The path where the layer is stored
+    pub store_path: String,
+}
+
+/// `ImageMetadata` struct holds information's about an image in the state.
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct ImageMetadata {
+    /// The image id
+    pub id: String,
+    /// The image reference, such as `docker.io/amd64/ubuntu`
+    pub reference: String,
+    /// The image digest
+    pub digest: String,
+    /// The image layers metas
+    pub layers: Vec<LayerMetadata>,
+    /// The image configuration
+    pub config: ImageConfiguration,
+}
+
+/// `State` is responsible about storing information's about images and layers.
+#[derive(Clone, Default, Deserialize, Debug, Serialize, PartialEq)]
+pub(crate) struct State {
+    /// The path to the state file
+    #[serde(skip)]
+    pub state_file: PathBuf,
+    /// An hashmap that holds every images pulled
+    images: HashMap<String, ImageMetadata>,
+    /// An hashmap that holds every layers pulled
+    layers: HashMap<String, LayerMetadata>,
+    /// An hashmap that holds snapshots with working directory index
+    snapshots: HashMap<String, usize>,
+}
+
+impl State {
+    /// Save the state into the file.
+    pub fn save(&self) -> crate::Result<()> {
+        let serialized = serde_json::to_string_pretty(&self)
+            .map_err(|e| Error::SerializeState(e.to_string()))?;
+
+        OpenOptions::new()
+            .write(true)
+            .open(&self.state_file)
+            .map_err(|e| Error::OpenStateFile(e.to_string()))?
+            .write_all(serialized.as_bytes())
+            .map_err(|e| Error::WriteStateFile(e.to_string()))
+    }
+}
+
+impl MetadataManager for State {
+    fn add_image(&mut self, image: &ImageMetadata) -> &mut State {
+        self.images.insert(image.id.clone(), image.clone());
+        self
+    }
+
+    fn add_layer(&mut self, layer: &LayerMetadata) -> &mut State {
+        self.layers
+            .insert(layer.compressed_digest.clone(), layer.clone());
+        self
+    }
+
+    fn add_snapshot(&mut self, index: usize) -> &mut Self {
```

Comment:
I'm not really sure to understand this point. Instead of carrying a HashMap, the `State` should carry a `snapshot_index` which will be incremented each time a mount is done ?

Reply:
`snapshot_index()`would be a method that gives you the next available index. Whenever a some asks for a mount of image docker.io/library/ubuntu:latest, you get a new index and you use it to create your mount path (e.g . `/var/run/kaps/snapshots/<index>/`). That's the path where you're snapshotter will mount the container image into.

Does that make more sense?